atl-checker/src/algorithms/certain_zero/search_strategy/linear_constrained_phi.rs

```diff
@@ -0,0 +1,274 @@
+                            }
+                        } else {
+                            LinearConstrainedPhi::NonLinear
+                        };
+                    }
+                    BinaryOpKind::And => {
+                        let lhs_con = self.map_expr_to_constraints(lhs, negated);
+                        let rhs_con = self.map_expr_to_constraints(rhs, negated);
+                        return if !negated {
+                            LinearConstrainedPhi::And(Box::new(lhs_con), Box::new(rhs_con))
+                        } else {
+                            LinearConstrainedPhi::Or(Box::new(lhs_con), Box::new(rhs_con))
+                        };
+                    }
+                    BinaryOpKind::Or => {
+                        let lhs_con = self.map_expr_to_constraints(lhs, negated);
+                        let rhs_con = self.map_expr_to_constraints(rhs, negated);
+                        return if !negated {
+                            LinearConstrainedPhi::Or(Box::new(lhs_con), Box::new(rhs_con))
+                        } else {
+                            LinearConstrainedPhi::And(Box::new(lhs_con), Box::new(rhs_con))
+                        };
+                    }
+                    // P -> Q == not P v Q
+                    BinaryOpKind::Implication => {
+                        let lhs_con = self.map_expr_to_constraints(lhs, !negated);
+                        let rhs_con = self.map_expr_to_constraints(rhs, negated);
+                        return if !negated {
+                            LinearConstrainedPhi::Or(Box::new(lhs_con), Box::new(rhs_con))
+                        } else {
+                            LinearConstrainedPhi::And(Box::new(lhs_con), Box::new(rhs_con))
+                        };
+                    }
+                    _ => {}
+                }
+            }
+            ExprKind::Number(n) => {
+                // Think of != as XOR in this situation
+                return if (*n == 0) != negated {
+                    LinearConstrainedPhi::False
+                } else {
+                    LinearConstrainedPhi::True
+                };
+            }
+            // https://en.wikipedia.org/wiki/Conditioned_disjunction
+            // Q ? P : R == (Q -> P) and (not Q -> R) == (Q and P) or (not Q and R)
+            ExprKind::TernaryIf(q, p, r) => {
+                let q = self.map_expr_to_constraints(q, negated);
+                let not_q = q.negated();
+                let p = self.map_expr_to_constraints(p, negated);
+                let r = self.map_expr_to_constraints(r, negated);
+                return if !negated {
+                    LinearConstrainedPhi::Or(
+                        Box::new(LinearConstrainedPhi::And(Box::new(q), Box::new(p))),
+                        Box::new(LinearConstrainedPhi::And(Box::new(not_q), Box::new(r))),
+                    )
+                } else {
```

Comment:
I'm not sure that the ternary is correctly mapped in the  `else`  case. As I can see it, it becomes `Q ? P : R` == `(not q Or not p) and (not q Or not r)`

Reply:
If negated, then we want to produce `not (Q and P) or (not Q and R)` => `(not (Q and P)) and (not (not Q and R))` => `(not Q or not P) and (Q or not R)`. And all terms are negated since negated was passed to it, so ultimately we want `(Q or P) and (not Q or R)`. So looks like I should swap the `q` and `not_q`. Heh.